workspace_tools/upload_results.py

```diff
@@ -68,28 +68,102 @@ def abort_build(args):
+
+    add_report(project_run_data, args.build_report, True, args.build_id, args.host_os)
+
+    if (args.test_report):
+        add_report(project_run_data, args.test_report, False, args.build_id, args.host_os)
+
+
+    ts_data = format_project_run_data(project_run_data)
+    r = requests.post(urlparse.urljoin(args.url, "api/projectRuns"), headers=create_headers(args), json=ts_data)
+    finish_command('add-project-runs', r)
+
+def format_project_run_data(project_run_data):
+    ts_data = {}
+    ts_data['projectRuns'] = []
+
+    for hostOs in project_run_data['projectRuns'].keys():
+        for platform in project_run_data['projectRuns'][hostOs].keys():
+            for toolchain in project_run_data['projectRuns'][hostOs][platform].keys():
+                for project in project_run_data['projectRuns'][hostOs][platform][toolchain].keys():
+                    ts_data['projectRuns'].append(project_run_data['projectRuns'][hostOs][platform][toolchain][project])
+
+    ts_data['platforms'] = list(project_run_data['platforms_set'])
+    ts_data['toolchains'] = list(project_run_data['toolchains_set'])
+    ts_data['names'] = list(project_run_data['names_set'])
+    ts_data['hostOses'] = list(project_run_data['hostOses_set'])
+
+    return ts_data
+
+# prd - Project Run Data - Datastructure containing all project runs
+# pt - Project Run - Data specific to one project run
+def project_run_exists_full(prd, pr):
+    if prd['projectRuns'][pr['hostOs']]:
+        if prd['projectRuns'][pr['hostOs']][pr['platform']]:
+            if prd['projectRuns'][pr['hostOs']][pr['platform']][pr['toolchain']]:
+                if prd['projectRuns'][pr['hostOs']][pr['platform']][pr['toolchain']][pr['project']]:
+                    return True
+
+    return False
+
+# prd - Project Run Data - Datastructure containing all project runs
+# pt - Project Run - Data specific to one project run
+def project_run_exists(prd, pr):
+    if pr['project'] in prd['projectRuns'][pr['hostOs']][pr['platform']][pr['toolchain']]:
+        return True
+    else:
+        return False
+
+# prd - Project Run Data - Datastructure containing all project runs
+# pt - Project Run - Data specific to one project run
+def prep_project_run(prd, pr):
+    if not pr['hostOs'] in prd['projectRuns']:
+        prd['projectRuns'][pr['hostOs']] = {}
+
+    if not pr['platform'] in prd['projectRuns'][pr['hostOs']]:
+        prd['projectRuns'][pr['hostOs']][pr['platform']] = {}
+
+    if not pr['toolchain'] in prd['projectRuns'][pr['hostOs']][pr['platform']]:
+        prd['projectRuns'][pr['hostOs']][pr['platform']][pr['toolchain']] = {}
+
+# prd - Project Run Data - Datastructure containing all project runs
+# pt - Project Run - Data specific to one project run
+def update_project_run(prd, pr):
+    prep_project_run(prd, pr)
+
+    if project_run_exists(prd, pr):
+        prd['projectRuns'][pr['hostOs']][pr['platform']][pr['toolchain']][pr['project']]['pass'] = pr['pass']
```

Comment:
Can we do anything to simplify the structure or at least describe it somewhere? This function could accept prd['projectRuns'][pr['hostOs']][pr['platform']][pr['toolchain']][pr['project']] as an argument ?

```
Hard to read?
prd['projectRuns'][pr['hostOs']][pr['platform']][pr['toolchain']][pr['project']]['pass']
```